File: 백준/JavaScript/유니온파인드/트리.js
const readline = require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
});

let input = [];
let t = 1;

function find(a) {
    if (a === parent[a]) {
        return a;
    }
    parent[a] = find(parent[a]);
    return parent[a];
}

function union(a, b) {
    const pa = find(a);
    const pb = find(b);

    if (pa < pb) {
        parent[pb] = pa;
    } else {
        parent[pa] = pb;
    }
}

rl.on("line", function (line) {
    input.push(line.trim());
}).on("close", function () {
    let index = 0;
    
    while (index < input.length) {
        let [n, m] = input[index].split(" ").map(Number);
        index++;

        if (n === 0 && m === 0) {
            break;
        }

        parent = Array.from({ length: n + 1 }, (_, i) => i);
        let cycle = new Set();

        for (let i = 0; i < m; i++) {
            let [a, b] = input[index].split(" ").map(Number);
            index++;

            if (find(a) === find(b)) {
                cycle.add(parent[a]);
            }

            if (cycle.has(parent[a]) || cycle.has(parent[b])) {
                cycle.add(parent[a]);
                cycle.add(parent[b]);
            }

            union(a, b);
        }

        for (let i = 1; i <= n; i++) {
            find(i); // 경로 압축
        }

        let s_parent = Array.from(new Set(parent.slice(1)));
        let treeCount = s_parent.reduce((count, p) => count + (cycle.has(p) ? 0 : 1), 0);

        if (treeCount === 0) {
            console.log(`Case ${t}: No trees.`);
        } else if (treeCount === 1) {
            console.log(`Case ${t}: There is one tree.`);
        } else {
            console.log(`Case ${t}: A forest of ${treeCount} trees.`);
        }
        t++;
    }
    process.exit();
});

File: 백준/Python/유니온파인드/Gold/트리.py
import sys
input = sys.stdin.readline

def find(a):
    if a == parent[a]:
        return a
    parent[a] = find(parent[a])
    return parent[a]

def union(a, b):
    pa = find(a)
    pb = find(b)

    if pa < pb:
        parent[pb] = pa
    else:
        parent[pa] = pb

t = 1
while True:
    n, m = map(int, input().split())
    if n == 0 and m == 0:
        break

    parent = [i for i in range(n + 1)]
    cycle = set()

    for _ in range(m):
        a, b = map(int, input().split())
        if find(a) == find(b):
            cycle.add(parent[a])

        if parent[a] in cycle or parent[b] in cycle:
            cycle.add(parent[a])
            cycle.add(parent[b])

        union(a, b)

    for i in range(1, n + 1):
        find(i)

    s_parent = list(set(parent))
    _sum = sum([1 if i not in cycle else 0 for i in s_parent]) - 1

    if _sum == 0:
        print(f'Case {t}: No trees.')
    elif _sum == 1:
        print(f'Case {t}: There is one tree.')
    else:
        print(f'Case {t}: A forest of {_sum} trees.')
    t += 1